### PR TITLE
feat(n3): StoreFactory

### PR DESCRIPTION
**Pull request by tpluscode**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/rdfjs/N3.js/pull/398/files#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556R32>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

File: types/n3/index.d.ts [hunk 1]
```diff
@@ -246,6 +246,10 @@ export class StreamWriter<Q extends RDF.BaseQuad = RDF.Quad> extends stream.Tran
     import(stream: RDF.Stream<Q>): EventEmitter;
 }
```
Comment (jeswr) [inline types/n3/index.d.ts:248]: I think this should be parameterised similar to the Store to permit the use of RDF.BaseQuad. It think it should also be permitted to supply BaseQuad in the constructor.

**Comment by jeswr** [comment]
@tpluscode whilst you're doing this could you also PR the type updates that indicates the Store now supports the full Dataset interface

**Comment by tpluscode** [comment]
> whilst you're doing this could you also PR the type updates that indicates the Store now supports the full Dataset interface.

You mean to replace `RDF.DatasetCore<OutQuad, InQuad>` with `RDF.Dataset<OutQuad, InQuad>`? It causes a lot of type issues which I don't really have the bandwidth now to pursue...

File: types/n3/index.d.ts (hunk 1 above)
Comment (tpluscode) [reply types/n3/index.d.ts:248]: Done

**Comment by jeswr** [comment]
> > whilst you're doing this could you also PR the type updates that indicates the Store now supports the full Dataset interface.
> 
> You mean to replace `RDF.DatasetCore<OutQuad, InQuad>` with `RDF.Dataset<OutQuad, InQuad>`? It causes a lot of type issues which I don't really have the bandwidth now to pursue...

Check, thanks for trying!

**Review by jeswr — APPROVED** [review]

**Comment by tpluscode** [comment]
Ready to merge

**Pull request merged**